# Genotyping SVs called using long-read data with BayesTyper 
BayesTyper is a powerful baysian based genotyping program. However, one draw back is that it is computationally intensive and has multiple intermediate steps to navigate. For more details on troubleshooting recommendations, please see [08_manta_bayestyper.md](https://github.com/janawold1/2022_MER_Omics_special_issue/blob/main/08_manta_bayestyper.md).

To begin, global variables were defined as below:
```
deepV=DEEPVARIANT:/kakapo-data/bayestyper/Trained_SV_scaffolds_renamed.sorted.vcf
out=/kakapo-data/bayestyper/
cute=/kakapo-data/bayestyper/cuteSV/
sniff=/kakapo-data/bayestyper/sniffles/q
ref=/kakapo-data/References/kakapo_autosomes.fa
trio=/kakapo-data/metadata/sample_trios.csv
```
See [09_ONT_SV_discovery.md](https://github.com/janawold1/2022_MER_Omics_special_issue/blob/main/09_ONT_SV_discovery.md) for more details on how SV calls using CuteSV and Sniffles were filtered prior to genotyping.

BayesTyper works best when SNPs are included. For this, the DeepVariant SNP call set available under the kākāpō125+ Sequencing Consortium were used. As outlined in [09_ONT_SV_discovery.md](https://github.com/janawold1/2022_MER_Omics_special_issue/blob/main/09_ONT_SV_discovery.md) all chromosome sizes were used to match kākāpō125+ and NCBI chromosome IDs.

## Running KMC and makeBloom
The third party program, KMC, and bayesTyperTools makeBloom are necessary to run BayesTyper cluster and BayesTyper genotype. It is independent of preparing the variant VCFs and can be done concurrrently. This was completed in [08_manta_bayestyper.md](https://github.com/janawold1/2022_MER_Omics_special_issue/blob/main/08_manta_bayestyper.md)

## Preparing SV candidates for genotyping
Unlike the short-read SV call set generated by Manta, the long-read call sets already included sequence information and were not only represented by symbolic alleles. Therefore, the conversion from symbolic alleles to sequences was not necessary.

To combine the VCFs generated by CuteSV and Sniffles with the SNPs called in the kākāpō125+ data set, we first normalised the SV calls.
```
bcftools norm --threads 24 -f ${ref} -m -any \
    -O z -o ${cute}02_bayestyper_candidates_norm.vcf.gz \
    ${cute}01_bayestyper_candidates.vcf

bcftools norm --threads 24 -f ${ref} -m -any \
    -O z -o ${sniff}02_bayestyper_candidates_norm.vcf.gz \
    ${sniff}01_bayestyper_candidates.vcf
```
 The SV calls and SNP calls were then merged as per:
```
bayesTyperTools combine \
    -v ${deepV},CuteSV:${cute}bayestyper_candidates_norm.vcf.gz \
    -o ${cute}03)combined_variants -z

bayesTyperTools combine \
    -v ${deepV},CuteSV:${sniff}bayestyper_candidates_norm.vcf.gz \
    -o ${sniff}03)combined_variants -z
```
However, not all SVs were included once this `bayesTyperTools combine` step was complete. To find which SVs were dropped from the `03_combined_variants` we first extracted the sites associated with either CuteSV or Sniffles.
```
bcftools view --threads 24 -i 'ACO="CuteSV"' \
    -O z -o ${cute}missing_sites/combined_cuteSV_variants.vcf.gz \
    ${cute}03_combined_variants.vcf.gz
bcftools view --threads 24 -i 'ACO="SNIFFLES"' \
    -O z -o ${sniff}missing_sites/combined_sniffles_variants.vcf.gz \
    ${sniff}03_combined_variants.vcf.gz
```
This indicated that 2,059 out of 2,238 CuteSV calls were included in the `03_combined_variants.vcf.gz` file and 5,030 out of 5,229 Sniffles calls were included for genotyping.

To find the type and location of the 'missing' SVs, we used `bcftools isec`.
```
bcftools isec \
    ${cute}02_bayestyper_candidates_norm.vcf.gz \
    ${cute}missing_sites/combined_cuteSV_variants.vcf.gz \
    -p ${cute}missing_sites/

bcftools isec \
    ${sniff}02_bayestyper_candidates_norm.vcf.gz \
    ${sniff}missing_sites/combined_sniffles_variants.vcf.gz \
    -p ${sniff}missing_sites/
```
It is notable that although 179 variants in the `03_combined_variants.vcf.gz` file for CuteSV and 199 of Sniffles varints appeared to be excluded following `bayesTyperTools combine`. However, more SVs failed to intersect with the normalised candidate file `02_bayestyper_candidates_norm.vcf.gz` for both SV discovery tools. A summary of these SVs is below.

|   SV type   | CuteSV | Sniffles |
|:-----------:|:------:|:--------:|
|   Deletion  |   66   |    87    |
| Duplication |  138   |    61    |
|  Insertion  |   26   |   179    |
|  Inversion  |   12   |    95    |
|    Total    |  242   |   422    |

Closely examining the number of SVs unique to `03_combined_variants.vcf.gz` confirmed that 63 CuteSV calls and 223 Sniffles calls had their ending location shifted during `bayesTyperTools combine`.

These were found and and summarised as per below.
```
bcftools query -f '%CHROM\t%POS\n' \
    ${cute}missing_sites/0001.vcf > ${cute}missing_sites/missing.txt
bcftools query -f '%CHROM\t%POS\n' \
    ${sniff}missing_sites/0001.vcf > ${sniff}missing_sites/missing.txt

bcftools query -R ${cute}missing_sites/missing.txt \
    -f '%SVTYPE\n' \
    ${cute}02_bayestyper_candidates_norm.vcf.gz | sort | uniq -c
bcftools query -R ${cute}missing_sites/missing.txt \
    -f '%SVTYPE\n' \
    ${cute}02_bayestyper_candidates_norm.vcf.gz | sort | uniq -c
```
The counts of shifted SVs for CuteSV and Sniffles are:
|   SV type   | CuteSV | Sniffles |
|:-----------:|:------:|:--------:|
|   Deletion  |   47   |   103    |
| Duplication |   24   |     6    |
|  Insertion  |   26   |   162    |
|  Inversion  |    0   |     0    |
|    Total    |   97   |   271    |

These details were used to resolve SV types after genotyping.

## Genotyping Variants
As noted previously, all unplaced scaffolds where ploidy was unknown were already excluded from candidate VCFs, BAM files and the reference. Note, no more than 30 indivs can be genotyped at once. The <samples>.tsv file should contain one sample per row with columns:
<sample_id>, <sex> and <kmc_output_prefix> and no header.

```
for samps in ${out}sample_batches/sample_batch*.tsv
    do
    base=$(basename ${samps} .tsv)
    bayesTyper cluster \
        --variant-file ${cute}03_combined_variants.vcf.gz \
        --samples-file ${samps} --genome-file ${ref} \
        --output-prefix ${cute}clusters/${base} --threads 24
    bayesTyper cluster \
        --variant-file ${sniff}03_combined_variants.vcf.gz \
        --samples-file ${samps} --genome-file ${ref} \
        --output-prefix ${sniff}clusters/${base} --threads 24
    bayesTyper genotype \
        --variant-clusters-file ${cute}clusters/${base}_unit_1/variant_clusters.bin \
        --cluster-data-dir ${cute}clusters/${base}_cluster_data/ \
        --samples-file ${samps} --genome-file ${ref} \
        --output-prefix ${cute}genotypes/${base}_genotypes \
        --threads 24
    bayesTyper genotype \
        --variant-clusters-file ${sniff}clusters/${base}_unit_1/variant_clusters.bin \
        --cluster-data-dir ${sniff}clusters/${base}_cluster_data/ \
        --samples-file ${samps} --genome-file ${ref} \
        --output-prefix ${sniff}genotypes/${base}_genotypes \
        --threads 24
done
```
## Merging SV genotype batches
To speed up the merging of genotype batches, the SV genotypes from Bayestyper were first extracted from the raw file and indexed with tabix.
```

for vcf in ${cute}genotypes/sample_batch{1..6}/*_genotypes.vcf
    do
    batch=$(basename $vcf _genotypes.vcf)
    echo "Extracting cuteSVs SV calls for $batch..."
    bcftools view --threads 64 -i 'ACO="CuteSV"' \
        -O z -o ${cute}genotypes/${batch}_cuteSV_genotypes.vcf.gz \
        $vcf
    echo "Now indexing $batch..."
    tabix ${cute}genotypes/${batch}_genotypes.vcf.gz
done &
for vcf in ${sniff}genotypes/sample_batch{1..6}/*_genotypes.vcf
    do
    batch=$(basename $vcf _genotypes.vcf)
   echo "Extracting Sniffles SV calls for $batch..."
    bcftools view --threads 64 -i 'ACO="SNIFFLES"' \
        -O z -o ${sniff}genotypes/${batch}_sniffles_genotypes.vcf.gz \
        $vcf
    echo "Now indexing $batch..."
    tabix ${sniff}genotypes/${batch}_genotypes.vcf.gz
done
wait
```
Then sample batches were merged according to sample ID with bcftools.
```
bcftools merge -m id -O z -o ${cute}04_raw_cuteSV_genotypes.vcf.gz \
    ${cute}genotypes/sample_batch*_cuteSV_genotypes.vcf.gz

bcftools merge -m id -O z -o ${sniff}/04_raw_sniffles_genotypes.vcf.gz \
    ${sniff}genotypes/sample_batch*_sniffles_genotypes.vcf.gz

tabix ${cute}04_raw_cuteSV_genotypes.vcf.gz
tabix ${sniff}04_raw_sniffles_genotypes.vcf.gz
```
This raw genotype file contains genotypes for SVs called by either CuteSV or Sniffles. However, the SV class is not included in this output. To interpret the genotype results from BayesTyper, we have to first reconcile the location of SVs in the genotypes with the location in the SV filtered call sets.  

## Finding SV type

As mentioned above in the *Preparing SV candidates for genotyping* section, BayesTyper removes symbolic alleles and can shift the variants location when running `bayesTyperTools combine`. However, this shifting in variant location can happen again during `bayesTyper cluster` and/or `bayesTyper genotype`.  

When attempting to reconcile the new locations of candidate SVs in the combined file we found 242 in the CuteSV call set and 422 in the Sniffles call set that had been shifted. Examining the overlap between the genotyped file (`04_raw_{cuteSV,sniffles}_genotypes.vcf.gz`) and SVs included in the combined variant file (`03_combined_variants.vcf.gz`) demonstrated that X CuteSV variants and 639 Sniffles variants had been shifted again.  
```
bcftools isec \
    ${cute}combined_cuteSV_variants.vcf.gz \
    ${cute}04_raw_cuteSV_genotypes.vcf.gz -p ${cute}
bcftools isec \
    ${sniff}combined_sniffles_variants.vcf.gz \
    ${sniff}04_raw_sniffles_genotypes.vcf.gz -p ${sniff}
```

To make comparisons among SV tools similar, the SV type called by either CuteSV or Sniffles were resoved as per below.

1. Examine overlap between SVs in `03_combined_variants.vcf.gz` and the locations of genotyped variants in `04_raw_{cuteSV,sniffles}_genotypes.vcf.gz`

```
bcftools query -f '%CHROM\t%POS\n' ${cute}04_raw_cuteSV_genotypes.vcf.gz > ${cute}geno_sites
bcftools query -f '%CHROM\t%POS\n' ${sniff}04_raw_sniffles_genotypes.vcf.gz > ${sniff}geno_sites
```

2. Count overlaping SVtypes
For each of the genotypes outputs we assessed the number of SVs overlapping with genotype outputs.  
```
bcftools query -T ${cute}geno_sites -f '%CHROM\t%POS\t%SVTYPE\t%SVLEN\n' \
    ${cute}02_bayestyper_candidates_norm.vcf.gz > ${cute}geno_types

bcftools query -T ${sniff}geno_sites -f '%CHROM\t%POS\t%SVTYPE\t%SVLEN\n' \
    ${sniff}02_bayestyper_candidates_norm.vcf.gz > ${sniff}geno_types
```
Once information about the type and size of SVs overlapping in the genotype output and input files was resolved, we ensured the number of SVs unique SVs matched.  
```
awk '{print $1, $2}' ${cute}geno_types | sort | uniq -c | sort -n | wc -l
awk '{print $1, $2}' ${sniff}geno_types | sort | uniq -c | sort -n | wc -l
```
We expected to find 2,058 SVs in the CuteSV call set and 5,030 SV in the Sniffles call set.  

A number of SVs from the VCF prior to running `bayesTyperTools combine` overlapped with the sites used in genotyping. This led to retrieving we couldn't resolve at least 15 more SVs from the CuteSV data call (n = 2,066) set and 37 more SVs from the Sniffles call set than expected (n = 5,067).

3. Find non-overlapping sites

The size of each of the overlapping sites were estimated as per below. This aided in resolving which of the overlapping SVs were represented in the genotype output.
```
bcftools query -t $site -f '%CHROM\t%POS\t%SVTYPE\t%SVLEN\n' \
    ${cute}02_bayestyper_candidates_norm.vcf.gz | wc
bcftools query -t $site -f '%ALT\n' \
    ${cute}04_raw_cuteSV_genotypes.vcf.gz | wc

bcftools query -t $site -f '%CHROM\t%POS\t%SVTYPE\t%SVLEN\n' \
    ${sniff}02_bayestyper_candidates_norm.vcf.gz | wc
bcftools query -t $site -f '%ALT\n' \
    ${sniff}04_raw_sniffles_genotypes.vcf.gz | wc
```

4. Annotate VCF
Once the `geno_types` file was edited for both the CuteSV and Sniffles call sets, the raw genotype files were annotated with SV type and length.  

To begin, the `geno_types` file was edited so the first line contained `#CHROM POS SVTYPE  SVLEN` with nano.  

And a `annots.hdr` file was created capturing the information to be appended into the header of the VCF. For example:
```
##INFO=<ID=SVTYPE,Number=1,Type=String,Description="Type of structural variant called by Sniffles">
##INFO=<ID=SVLEN,Number=1,Type=String,Description="Length of structural variant called by Sniffles">
```
Now time to annotate the file to continue with the genotype based analyses.
```
bgzip ${cute}geno_types
bgzip ${sniff}geno_types

tabix -s1 -b2 -e2 ${cute}geno_types.gz
tabix -s1 -b2 -e2 ${sniff}geno_types.gz

bcftools annotate -a ${cute}geno_types.gz -h ${cute}annots.hdr \
    -c CHROM,POS,SVTYPE,SVLEN -O v -o ${cute}05_cuteSV_annotated.vcf \
    ${cute}04_raw_cuteSV_genotypes.vcf

bcftools annotate -a ${sniff}geno_types.gz -h ${sniff}annots.hdr \
    -c CHROM,POS,SVTYPE,SVLEN -O v -o ${sniff}05_sniffles_annotated.vcf \
    ${sniff}04_sniffles_genotypes.vcf.gz
```

## Genotype quality filtering

SVs were simultaneously filtered for missingness and to ensure variant sites were included. As with the Manta SV call sets, for all sites where at least one individual carried an alternate allele (`N_PASS(GT=="alt") >= 1`), we allowed an individual missingness of 20% (`N_PASS(GT=="mis") < 35`) and required that at least 80% of all samples passed all four of BayesTyper's 'hard' filtering thresholds (`N_PASS(SAF!=0) < 35`).
```
 bcftools view --threads 24 \
    -i '((N_PASS(GT=="mis") < 35) & (N_PASS(GT=="alt") >= 1) & (N_PASS(SAF!=0) < 35))' \
    -O z -o ${cute}06_cuteSV_genotype_filtered.vcf.gz \
    ${cute}05_raw_cuteSV_annotated.vcf

 bcftools view --threads 24 \
    -i '((N_PASS(GT=="mis") < 35) & (N_PASS(GT=="alt") >=1 ) & (N_PASS(SAF!=0) < 35))' \
    -O z -o ${sniff}06_sniffles_genotype_filtered.vcf.gz \
    ${sniff}05_sniffles_annotated.vcf
```
We are now ready to test Mendelian trios and prepare summary files.

## Mendelian Inheritance Tests

To test the consistency of genotyping results, tests of mendelian inheritance were used. Sites that conformed to mendelian expectations for 120 trios were found with:
```
bcftools +mendelian -m a -T ${trio} -O v \
    -o ${cute}07_cuteSV_genotype_filtered_trios.vcf \
    ${cute}06_cuteSV_genotype_filtered.vcf.gz

bcftools +mendelian -m a -T ${trio} -O v \
    -o ${sniff}07_sniffles_genotype_filtered_trios.vcf \
    ${sniff}06_sniffles_genotype_filtered.vcf
```
The proportion of genotyped sites that adhered to Mendelian inheritance expectations were found with:
```
for i in {0,0.05,0.1,0.2}
    do
    bcftools query -i "(MERR / N_PASS(GT!="mis")) <=${i}" \
        -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t${i}_fail_cuteSV_genofilter\n' \
        ${cute}07_trios.vcf >> /kakapo-data/bayestyper/summary/cuteSV_mendel.tsv
    bcftools query -i "(MERR / N_PASS(GT!="mis")) <=${i}" \
        -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t${i}_fail_sniffles_genofilter\n' \
        ${sniff}08_sniffles_trios.vcf >> /kakapo-data/bayestyper/summary/sniffles_mendel.tsv
done
```
## Summarising the number of SVs carried by individuals and generations  
A summary of individual genotypes and generation for the subset of individuals that are not resulting from backcrossing (i.e., true founders, F1 and F2's) were used to assess the proportion of SVs passing different Mendelian Inheritance thresholds.  
```
bcftools view \
    -S /kakapo-data/metadata/target_samples \
    -O v -o ${cute}08_cuteSV_genofilter_trio_target_samples.vcf \
    ${cute}07_cuteSV_genotype_filtered_trios.vcf
bcftools view \
    -S /kakapo-data/metadata/target_samples \
    -O v -o ${cute}08_cuteSV_genofilter_trio_target_samples.vcf \
    ${cute}07_cuteSV_genotype_filtered_trios.vcf

while read -r line
    do
    indiv=$(echo $line | awk '{print $1}')
    gen=$(echo $line | awk '{print $2}')
    echo "Counting SVs for $indiv..."
    bcftools view -s ${indiv} ${cute}08_cuteSV_genotypefiltered_trio_samples.vcf \
        | bcftools query -i 'GT!= "RR" & GT!="mis"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\n' \
        | awk -v var="$gen" '{print $0"\t"var"\tCuteSV"}' >> ${out}bayestyper/summary/cute_generations.tsv
    bcftools view -s ${indiv} ${sniff}08_sniffles_genotypefiltered_trio_target_samples.vcf \
        | bcftools query -i 'GT!= "RR" & GT!="mis"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\n' \
        | awk -v var="$gen" '{print $0"\t"var"\tSniffles"}' >> ${out}bayestyper/summary/sniffles_generations.tsv
done < /kakapo-data/metadata/generations.tsv
```