# Genotyping SVs called using long-read data with BayesTyper 
BayesTyper is a powerful baysian based genotyping program. However, one draw back is that it is computationally intensive and has multiple intermediate steps to navigate. For more details on troubleshooting recommendations, please see [08_manta_bayestyper.md](https://github.com/janawold1/2022_MER_Omics_special_issue/blob/main/08_manta_bayestyper.md).

To begin, global variables were defined as below:
```
deepV=DEEPVARIANT:/kakapo-data/bayestyper/Trained_SV_scaffolds_renamed.sorted.vcf
out=/kakapo-data/bayestyper/
cute=/kakapo-data/bayestyper/cuteSV/
sniff=/kakapo-data/bayestyper/sniffles/q
ref=/kakapo-data/References/kakapo_autosomes.fa
trio=/kakapo-data/metadata/sample_trios.csv
```
See [09_ONT_SV_discovery.md](https://github.com/janawold1/2022_MER_Omics_special_issue/blob/main/09_ONT_SV_discovery.md) for more details on how SV calls using CuteSV and Sniffles were filtered prior to genotyping.

BayesTyper works best when SNPs are included. For this, the DeepVariant SNP call set available under the kākāpō125+ Sequencing Consortium were used. As outlined in [09_ONT_SV_discovery.md](https://github.com/janawold1/2022_MER_Omics_special_issue/blob/main/09_ONT_SV_discovery.md) all chromosome sizes were used to match kākāpō125+ and NCBI chromosome IDs.

## Convert VCFs to remove symbolic alleles
This converts all symbolic alleles to a full sequence. The conversion results in large VCFs, so ensure that the output is gzipped with the -z flag. Note, bayesTyper does NOT support bgzip files. BayesTyper benefits from including SNP data as it helps with initialising the kmers. Therefore it's recommended to conduct SNP discovery for inclusion in SV genotyping.

Unlike the short-read SV call set generated by Manta, the long-read call sets already included sequence information and were not only represented by symbolic alleles. Therefore, this conversion was not needed for these VCFs.

To combine the VCFs generated by CuteSV and Sniffles with the SNPs called in the kākāpō125+ data set, we first normalised the SV calls.
```
bcftools norm --threads 24 -f ${ref} -m -any \
    -O z -o ${cute}bayestyper_candidates_norm.vcf.gz \
    ${cute}bayestyper_candidates.vcf

bcftools norm --threads 24 -f ${ref} -m -any \
    -O z -o ${sniff}bayestyper_candidates_norm.vcf.gz \
    ${sniff}bayestyper_candidates.vcf
```
 The SV calls and SNP calls were then merged as per:
```
bayesTyperTools combine \
    -v ${deepV},CuteSV:${cute}bayestyper_candidates_norm.vcf.gz \
    -o ${cute}combined_variants -z

bayesTyperTools combine \
    -v ${deepV},CuteSV:${sniff}bayestyper_candidates_norm.vcf.gz \
    -o ${sniff}combined_variants -z
```

## Running KMC and makeBloom
The third party program, KMC, and bayesTyperTools makeBloom are necessary to run BayesTyper cluster and BayesTyper genotype. It is independent of preparing the variant VCFs and can be done concurrrently.

It is really easy to over-resource KMC, that is to say if you provide it with too
much memory/threads the program freezes. I have had the best luck with 24Gb RAM and 48 threads as per below. It also found that I needed to run KMC one individual at a time. I had a lot of isls when I attempted to run multiple individuals at once since KMC uses the same temp file naming convention each time it runs. Basically temp files from different runs were overwriting each other and I needed to designate different locations for the temp files.
```
ulimit -n 2048
for mkmc in ${male}batch{01..14}/*_nodup.bam
    do
    indiv=$(basename ${mkmc} _nodup.bam)
    echo "Running KMC for ${indiv}..."
    kmc -k55 -ci1 -m24 -t48 -fbam ${mkmc} ${out}kmc/${indiv}_KMC.res ${out}kmc_tmp/
    echo "Running makeBloom for ${indiv}..."
    bayesTyperTools makeBloom -k ${out}kmc/${indiv}_KMC.res -p 8
done
for fkmc in ${female}batch{01..11}/*_nodup.bam
    do
    indiv=$(basename ${fkmc} _nodup.bam)
    echo "Running KMC for ${indiv}..."
    kmc -k55 -ci1 -m24 -t48 -fbam ${fkmc} ${out}kmc/${indiv}_KMC.res ${out}kmc_tmp/
    echo "Running makeBloom for ${indiv}..."
    bayesTyperTools makeBloom -k ${out}kmc/${indiv}_KMC.res -p 8
done
```
As noted previously, all unplaced scaffolds where ploidy was unknown were already excluded from candidate VCFs, BAM files and the reference. Note, no more than 30 indivs can be genotyped at once. The <samples>.tsv file should contain one sample per row with columns:
<sample_id>, <sex> and <kmc_output_prefix> and no header.

```
for samps in ${out}sample_batches/sample_batch*.tsv
    do
    base=$(basename ${samps} .tsv)
    bayesTyper cluster \
        --variant-file ${cute}combined_variants.vcf.gz \
        --samples-file ${samps} --genome-file ${ref} \
        --output-prefix ${cute}clusters/${base}/ --threads 24
    bayesTyper cluster \
        --variant-file ${sniff}combined_variants.vcf.gz \
        --samples-file ${samps} --genome-file ${ref} \
        --output-prefix ${sniff}clusters/${base}/ --threads 24
    bayesTyper genotype \
        --variant-clusters-file ${cute}clusters/${base}_unit_1/variant_clusters.bin \
        --cluster-data-dir ${cute}clusters/${base}_cluster_data/ \
        --samples-file ${samps} --genome-file ${ref} \
        --output-prefix ${cute}genotypes/${base}_genotypes \
        --threads 24 --chromosome-ploidy-file ${out}ploidy.tsv
    bayesTyper genotype \
        --variant-clusters-file ${sniff}clusters/${base}_unit_1/variant_clusters.bin \
        --cluster-data-dir ${sniff}clusters/${base}_cluster_data/ \
        --samples-file ${samps} --genome-file ${ref} \
        --output-prefix ${sniff}genotypes/${base}_genotypes \
        --threads 24 --chromosome-ploidy-file ${out}ploidy.tsv
done
```
## Merging SV genotype batches
```
for vcf in ${cute}genotypes/sample_batch{1..6}/*_genotypes.vcf
    do
    batch=$(basename $vcf _genotypes.vcf)
    bgzip ${vcf}
    echo "Now indexing $vcf"
    tabix ${cute}genotypes/${batch}/${batch}_genotypes.vcf.gz
done

bcftools merge -m id -O z -o joint_filtered/06_manta_genotypes.vcf.gz \
    joint_filtered/sample_batch1/sample_batch1_genotypes.vcf.gz \
    joint_filtered/sample_batch2/sample_batch2_genotypes.vcf.gz \
    joint_filtered/sample_batch3/sample_batch3_genotypes.vcf.gz \
    joint_filtered/sample_batch4/sample_batch4_genotypes.vcf.gz \
    joint_filtered/sample_batch5/sample_batch5_genotypes.vcf.gz \
    joint_filtered/sample_batch6/sample_batch6_genotypes.vcf.gz

 bcftools view --threads 24 -i '(ACO="MANTA") & ((N_PASS(GT=="mis") < 17) & (N_PASS(GT="alt")>=1))' \
    -O v -o joint_filtered/07_joint_filtered_genotypes.vcf \
    joint_filtered/06_joint_genotypes.vcf.gz

for vcf in ${out}bayestyper/batch_filtered/sample_batch{1..6}/*_genotypes.vcf.gz
    do
    batch=$(basename $vcf _genotypes.vcf.gz)
    echo "Converting $vcf to bgzip compression..."
    gunzip $vcf
    bgzip ${out}bayestyper/batch_filtered/${batch}/${batch}_genotypes.vcf
    echo "Now indexing $vcf"
    tabix ${out}bayestyper/batch_filtered/${batch}/${batch}_genotypes.vcf.gz
done

bcftools merge -m id -O z -o batch_filtered/06_manta_genotypes.vcf.gz \
    ${out}bayestyper/batch_filtered/sample_batch1/sample_batch1_genotypes.vcf.gz \
    ${out}bayestyper/batch_filtered/sample_batch2/sample_batch2_genotypes.vcf.gz \
    ${out}bayestyper/batch_filtered/sample_batch3/sample_batch3_genotypes.vcf.gz \
    ${out}bayestyper/batch_filtered/sample_batch4/sample_batch4_genotypes.vcf.gz \
    ${out}bayestyper/batch_filtered/sample_batch5/sample_batch5_genotypes.vcf.gz \
    ${out}bayestyper/batch_filtered/sample_batch6/sample_batch6_genotypes.vcf.gz

 bcftools view --threads 24 -i '(ACO="MANTA") & ((N_PASS(GT=="mis") < 17) & (N_PASS(GT="alt")>=1))' \
    -O v -o ${out}bayestyper/batch_filtered/07_batch_filtered_genotypes.vcf \
    ${out}bayestyper/batch_filtered/06_batch_genotypes.vcf.gz

```

## Finding SV type

BayesTyper removes symbolic alleles from the genotype files. To make comparisons among SV tools similar, the SV type called by Manta was resoved as per below.
1) First identify the locations of genotyped variants:
```
bcftools query -f '%CHROM\t%POS\n' ${out}bayestyper/batch_filtered/08_batch_filtered_trios.vcf > ${out}bayestyper/batch_genos
bcftools query -f '%CHROM\t%POS\n' ${out}bayestyper/joint_filtered/08_joint_filtered_trios.vcf > ${out}bayestyper/joint_genos
```

2) Count overlaping SVtypes:
```
bcftools query -T ${out}bayestyper/batch_genos -f '%SVTYPE\n' \
    ${out}bayestyper/batch_filtered/02_batch_filtered_norm.vcf.gz | sort | uniq -c

bcftools query -T ${out}bayestyper/joint_genos -f '%SVTYPE\n' \
    ${out}bayestyper/joint_filtered/02_joint_filtered_norm.vcf.gz | sort | uniq -c
```

Found that 76 SVs don't overlap in the batch data and 126 SVs don't overlap in the joint data. 

3) Find non-overlapping sites (i.e., those present in genotyped output, but absent in call set):
```

```

4) Annotate VCF for individual counts:
First prepare the annotation file by identifying the resolvable SVs and their types:
```
bcftools query -f '%CHROM\t%POS\n' batch_filtered/07_batch_filtered_genotypes.vcf > batch_geno_sites
bcftools query -T batch_geno_sites -f '%CHROM\t%POS\t%SVTYPE\t%SVLEN\n' batch_filtered/02_batch_filtered_norm.vcf.gz > batch_conversion

bcftools query -f '%CHROM\t%POS\n' joint_filtered/07_joint_filtered_genotypes.vcf > joint_geno_sites
bcftools query -T joint_geno_sites -f '%CHROM\t%POS\t%SVTYPE\t%SVLEN\n' joint_filtered/02_joint_filtered_norm.vcf.gz > joint_conversion
```

Then edit the ```batch_conversion``` file so the first line in this file contains ```#CHROM POS SVTYPE  SVLEN``` with nano. 

And create a file that captures the information to be appended into the header of the VCF.

Contents of annots.hdr for example:
```
##INFO=<ID=SVTYPE,Number=1,Type=String,Description="Type of structural variant called by Sniffles">
##INFO=<ID=SVLEN,Number=1,Type=String,Description="Length of structural variant called by Sniffles">
```
Now time to annotate the file to continue with the genotype based analyses.
```
bgzip batch_conversion
bgzip joint_conversion

tabix -s1 -b2 -e2 batch_conversion.gz
tabix -s1 -b2 -e2 joint_conversion.gz

bcftools annotate -a batch_conversion.gz -h annots.hdr \
    -c CHROM,POS,SVTYPE,SVLEN -O v -o 09_batch_annotated.vcf \
    ${out}bayestyper/batch_filtered/07_batch_filtered_genotypes.vcf

bcftools annotate -a joint_conversion.gz -h annots.hdr \
    -c CHROM,POS,SVTYPE,SVLEN -O v -o 09_joint_annotated.vcf \
    ${out}bayestyper/joint_filtered/07_joint_filtered_genotypes.vcf
```

## Mendelian Inheritance Tests
Tests of Mendelian Inheritance were conducted as per: 

```
bcftools +mendelian -m a -T ${trio} -O v -o ${out}bayestyper/joint_filtered/09_joint_annotated_trios.vcf \
    ${out}bayestyper/joint_filtered/08_joint_annotated.vcf
bcftools +mendelian -m a -T ${trio} -O v -o ${out}bayestyper/batch_filtered/09_batch_annotated_trios.vcf \
    ${out}bayestyper/batch_filtered/08_batch_annotated.vcf

bcftools query -i '(MERR / N_PASS(GT!="mis")) <=0' \
    -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t0_fail_joint_manta_genofilter\n' \
    ${out}bayestyper/joint_filtered/09_joint_annotated_trios.vcf >> ${out}bayestyper/summary/manta_mendel.tsv

bcftools query -i '(MERR / N_PASS(GT!="mis")) <=0.05' \
    -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t0.05_fail_joint_manta_genofilter\n' \
    ${out}bayestyper/joint_filtered/09_joint_annotated_trios.vcf >> ${out}bayestyper/summary/manta_mendel.tsv

bcftools query -i '(MERR / N_PASS(GT!="mis")) <=0.1' \
    -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t0.1_fail_joint_manta_genofilter\n' \
    ${out}bayestyper/joint_filtered/09_joint_annotated_trios.vcf >> ${out}bayestyper/summary/manta_mendel.tsv

bcftools query -i '(MERR / N_PASS(GT!="mis")) <=0.2' \
    -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t0.2_fail_joint_manta_genofilter\n' \
    ${out}bayestyper/joint_filtered/09_joint_annotated_trios.vcf >> ${out}bayestyper/summary/manta_mendel.tsv



bcftools query -i '(MERR / N_PASS(GT!="mis")) <=0' \
    -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t0_fail_batch_manta_genofilter\n' \
    ${out}bayestyper/batch_filtered/09_batch_annotated_trios.vcf >> ${out}bayestyper/summary/manta_mendel.tsv

bcftools query -i '(MERR / N_PASS(GT!="mis")) <=0.05' \
    -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t0.05_fail_batch_manta_genofilter\n' \
    ${out}bayestyper/batch_filtered/09_batch_annotated_trios.vcf >> ${out}bayestyper/summary/manta_mendel.tsv

bcftools query -i '(MERR / N_PASS(GT!="mis")) <=0.1' \
    -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t0.1_fail_batch_manta_genofilter\n' \
    ${out}bayestyper/batch_filtered/09_batch_annotated_trios.vcf >> ${out}bayestyper/summary/manta_mendel.tsv

bcftools query -i '(MERR / N_PASS(GT!="mis")) <=0.2' \
    -f '%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\t0.2_fail_batch_manta_genofilter\n' \
    ${out}bayestyper/batch_filtered/09_batch_annotated_trios.vcf >> ${out}bayestyper/summary/manta_mendel.tsv
```
## Summarising the number of SVs carried by individuals

```
while read -r line
    do
    indiv=$(echo $line | awk '{print $1}')
    gen=$(echo $line | awk '{print $2}')
    echo "Counting SVs for $indiv..."
    bcftools view -s ${indiv} ${out}bayestyper/batch_filtered/09_batch_annotated_trios.vcf | bcftools query -i 'GT!= "RR" & GT!="mis"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\n' | awk -v var="$gen" '{print $0"\t"var}' >> ${out}bayestyper/summary/manta_batch_generations.tsv
    bcftools view -s ${indiv} ${out}bayestyper/joint_filtered/09_joint_annotated_trios.vcf | bcftools query -i 'GT!= "RR" & GT!="mis"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVLEN\t%SVTYPE\n' | awk -v var="$gen" '{print $0"\t"var}' >> ${out}bayestyper/summary/manta_joint_generations.tsv
done < /kakapo-data/metadata/generations.tsv
```


## Lineage Comparisons
```
mkdir -p ${out}bayestyper/lineage_{batch,joint}_comparisons

bcftools view -s M /kakapo-data/bwa/manta/raw_variants/batch_total.vcf.gz | \
    bcftools view -i 'GT=="alt"' -O z -o ${out}bayestyper/lineage_batch_comparisons/unfiltered/RH_unfiltered_variants.vcf.gz
bcftools view -s M /kakapo-data/bwa/manta/raw_variants/joint_total.vcf.gz | \
    bcftools view -i 'GT=="alt"' -O z -o ${out}bayestyper/lineage_joint_comparisons/unfiltered/RH_unfiltered_variants.vcf.gz

bcftools view -s ^M,G,F,R,N,P,O,S /kakapo-data/bwa/manta/raw_variants/batch_total.vcf.gz |\
    bcftools view -i 'GT=="alt"' -O z -o ${out}bayestyper/lineage_batch_comparisons/unfiltered/SI_unfiltered_variants.vcf.gz
bcftools view -s ^M,G,F,R,N,P,O,S /kakapo-data/bwa/manta/raw_variants/joint_total.vcf.gz |\
    bcftools view -i 'GT=="alt"' -O z -o ${out}bayestyper/lineage_joint_comparisons/unfiltered/SI_unfiltered_variants.vcf.gz

bcftools index ${out}bayestyper/lineage_batch_comparisons/unfiltered/RH_unfiltered_variants.vcf.gz
bcftools index ${out}bayestyper/lineage_batch_comparisons/unfiltered/SI_unfiltered_variants.vcf.gz

bcftools index ${out}bayestyper/lineage_joint_comparisons/unfiltered/RH_unfiltered_variants.vcf.gz
bcftools index ${out}bayestyper/lineage_joint_comparisons/unfiltered/SI_unfiltered_variants.vcf.gz

bcftools isec ${out}bayestyper/lineage_batch_comparisons/unfiltered/RH_unfiltered_variants.vcf.gz \
    ${out}bayestyper/lineage_batch_comparisons/unfiltered/SI_unfiltered_variants.vcf.gz \
    -p ${out}bayestyper/lineage_batch_comparisons/unfiltered/

bcftools isec ${out}bayestyper/lineage_joint_comparisons/unfiltered/RH_unfiltered_variants.vcf.gz \
    ${out}bayestyper/lineage_joint_comparisons/unfiltered/SI_unfiltered_variants.vcf.gz \
    -p ${out}bayestyper/lineage_joint_comparisons/unfiltered/

bcftools query -f '%CHROM\t%POS\n' \
    ${out}bayestyper/lineage_batch_comparisons/unfiltered/0000.vcf > ${out}bayestyper/lineage_batch_comparisons/unfiltered/Fiordland_unfiltered_private_sites.txt
bcftools query -f '%CHROM\t%POS\n' \
    ${out}bayestyper/lineage_batch_comparisons/unfiltered/0001.vcf > ${out}bayestyper/lineage_batch_comparisons/unfiltered/Rakiura_unfiltered_private_sites.txt
bcftools query -f '%CHROM\t%POS\n' \
    ${out}bayestyper/lineage_batch_comparisons/unfiltered/0002.vcf > ${out}bayestyper/lineage_batch_comparisons/unfiltered/Shared_unfiltered_sites.txt

bcftools query -f '%CHROM\t%POS\n' \
    ${out}bayestyper/lineage_joint_comparisons/unfiltered/0000.vcf > ${out}bayestyper/lineage_joint_comparisons/unfiltered/Fiordland_unfiltered_private_sites.txt
bcftools query -f '%CHROM\t%POS\n' \
    ${out}bayestyper/lineage_joint_comparisons/unfiltered/0001.vcf > ${out}bayestyper/lineage_joint_comparisons/unfiltered/Rakiura_unfiltered_private_sites.txt
bcftools query -f '%CHROM\t%POS\n' \
    ${out}bayestyper/lineage_joint_comparisons/unfiltered/0002.vcf > ${out}bayestyper/lineage_joint_comparisons/unfiltered/Shared_unfiltered_sites.txt

while read -r line
    do
    indiv=$(echo $line | awk '{print $1}')
    gen=$(echo $line | awk '{print $2}')
    echo "Counting SVs for $indiv..."
    bcftools view -s ${indiv} -R ${out}bayestyper/lineage_batch_comparisons/unfiltered/Fiordland_unfiltered_private_sites.txt raw_variants/batch_total.vcf.gz | bcftools query -i 'GT=="alt"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVTYPE\tFiordland_unfiltered_lineage\n' >> ${out}mantaB_lineage_counts.tsv
    bcftools view -s ${indiv} -R ${out}bayestyper/lineage_batch_comparisons/unfiltered/Rakiura_unfiltered_private_sites.txt raw_variants/batch_total.vcf.gz | bcftools query -i 'GT=="alt"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVTYPE\tRakiura_unfiltered_lineage\n' >> ${out}mantaB_lineage_counts.tsv
    bcftools view -s ${indiv} -R ${out}bayestyper/lineage_batch_comparisons/unfiltered/Shared_unfiltered_sites.txt raw_variants/batch_total.vcf.gz | bcftools query -i 'GT=="alt"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVTYPE\tShared_unfiltered_lineage\n' >> ${out}mantaB_lineage_counts.tsv
done < /kakapo-data/metadata/generations.tsv

while read -r line
    do
    indiv=$(echo $line | awk '{print $1}')
    gen=$(echo $line | awk '{print $2}')
    echo "Counting SVs for $indiv..."
    bcftools view -s ${indiv} -R ${out}bayestyper/lineage_joint_comparisons/unfiltered/Fiordland_unfiltered_private_sites.txt ${out}raw_variants/joint_total.vcf.gz | bcftools query -i 'GT=="alt"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVTYPE\tFiordland_unfiltered_lineage\n' >> ${out}mantaJ_lineage_counts.tsv
    bcftools view -s ${indiv} -R ${out}bayestyper/lineage_joint_comparisons/unfiltered/Rakiura_unfiltered_private_sites.txt ${out}raw_variants/joint_total.vcf.gz | bcftools query -i 'GT=="alt"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVTYPE\tRakiura_unfiltered_lineage\n' >> ${out}mantaJ_lineage_counts.tsv
    bcftools view -s ${indiv} -R ${out}bayestyper/lineage_joint_comparisons/unfiltered/Shared_unfiltered_sites.txt ${out}raw_variants/joint_total.vcf.gz | bcftools query -i 'GT=="alt"' -f '[%SAMPLE]\t%CHROM\t%POS\t%END\t%SVTYPE\tShared_unfiltered_lineage\n' >> ${out}mantaJ_lineage_counts.tsv
done < /kakapo-data/metadata/generations.tsv

```