---
title: 'Characteristics of Structural Variants discovered using different tools in the critically endangered Kākāpō'
author: "Jana R. Wold"
date: "18 November 2022"
output: "pdf"
---

Here is the outline of charting differences among structural variants (SVs) identified using three short-read SV discovery tools, Delly, Manta and Smoove and two long-read SV discovery tools, CuteSV and Sniffles. SV calls for both long-read discovery tools were refined using Jasmine prior to applying SV quality filtering parameters. 

In this Rmd, we examine the variation in the number of SVs discovered by type, size and chromosome for each of these five tools. 
```{r setup, include=FALSE}
library(ggplot2)
library(ggridges)
library(ggpubr)
library(gridExtra)
library(dplyr)
library(tidyverse)
library(adegenet)
library(hierfstat)
library(vcfR)

#install.packages("devtools")
#devtools::install_github("G-Thomson/Manu")
library(Manu)

pal <- get_pal("Kakapo")
get_pal("Kakapo")
print_pal(pal)

knitr::opts_chunk$set(dev = c("svg", "png"),
                      dpi = 300,
                      echo = FALSE,
                      cache = TRUE)

setwd("G:/My Drive/Data/Kakapo/2022_MER_submission")
```

Three types of inputs for each of the five tools were generated: 1) a *_summary.tsv file (a.k.a Summary file); 2) a *_generations.tsv file (Generation file); and 3) a *_lineage_counts.tsv (Lineage file) file for each SV discovery tool/strategy. The summary file describes characteristics of SVs called by each tool and includes the headings:
| chrom | start | end | size | SV_type | data | chrom_size |
|:-----:|:-----:|:---:|:----:|:-------:|:----:|:----------:|
Here, and in all files described here, chrom represents the autosome the SV was discovered on, start denotes the starting position, while end denotes the ending position for the SV, and size represents the length of the SV in base pairs. SV_type represents the type of SV called (i.e., breakend, deletion, duplication, insertion, or inversion), while data indicates whether this SV is described in the call set filtered for SV quality and/or the unfiltered call set.

The generation file has a line for each variable SV genotyped in individuals. Like the summary file, it contains the chrom, start, end, size, SV_type and data fields. But it also includes an 'indiv' field denoting the individual carrying the SV and a 'gen' field that represents the individuals' lineage and generation.

# SV summaries
## Counts
Here I explored the number of structural variants per data set before and after filtering for SV quality and genotype quality. 

To begin we have structural variants discovered using the short-read SV tools Delly, Manta, and Smoove, and the long-read SV tools CuteSV and Sniffles. Due to the high computational requirements for Manta, two strategies were used for structural variant discovery, 1) a batched approach; and 2) joint variant calling.

The below plots were used to generate Figure 1 (SV filter overlaps) and Figure 2 (Chromosome counts and impacted basepairs). First the SV summary file for each SV discovery tool was imported, concatenated into one variable `svs` and split into filtered and unfiltered data sets.
```{r load type count data, echo=FALSE, fig.align='center'}
delly <-read.table("inputs/delly_summary.tsv", sep = "\t", header = TRUE)
manta <-read.table("inputs/manta_summary.tsv", sep = "\t", header = TRUE)
smoove <-read.table("inputs/smoove_summary.tsv", sep = "\t", header = TRUE)
cute  <- read.table("inputs/cute_summary.tsv", sep = "\t", header = TRUE)
sniff <- read.table("inputs/sniffles_summary.tsv", sep = "\t", header = TRUE)
svs <- rbind(delly, manta, smoove, cute, sniff)

unfiltered <- svs %>%
                filter(if_any(data, ~grepl('unfiltered' , .)))
filtered <- svs %>%
                filter(if_any(data, ~grepl('SVfiltered' , .)))
```

The overall type counts were then visualised for both the filtered and unfiltered data sets.

```{r plotting overall counts, echo=FALSE, fig.align='center'}
ggplot(unfiltered, aes(x=SV_type, fill=data)) +
  geom_histogram(stat = "count", position = position_dodge()) +
  labs(x = "Structural Variant Type", y = "Count", title = "Count of all Structural Variants") +
  scale_fill_manual(values = c("delly_unfiltered" = "#7D9D33", "mantaB_unfiltered" = "#CED38C", "mantaJ_unfiltered" = "#DCC949", "smoove_unfiltered" = "#BCA888", "cuteSV_unfiltered" = "#CD8862", "sniffles_unfiltered" = "#775B24")) +
  theme_light()

ggplot(filtered, aes(x=SV_type, fill=data)) +
  geom_histogram(stat = "count", position = position_dodge()) +
  labs(x = "Structural Variant Type", y = "Count", title = "Count of Filtered Structural Variants") +
  scale_fill_manual(values = c("delly_SVfiltered" = "#7D9D33", "mantaB_SVfiltered" = "#CED38C", "mantaJ_SVfiltered" = "#DCC949", "smoove_SVfiltered" = "#BCA888", "cuteSV_SVfiltered" = "#CD8862", "sniffles_SVfiltered" = "#775B24")) +
  theme_light()

#pdf("plots/Fig1_total_counts.pdf")
filtered %>%
  group_by(data) %>%
  ggplot(aes(x=data, fill=data)) +
    geom_histogram(stat = "count", position = position_dodge()) +
    labs(x = "Structural Variant Caller", y = "Count", title = "Count of Filtered Structural Variants") +
    scale_fill_manual(values = c("delly_SVfiltered" = "#7D9D33", "mantaB_SVfiltered" = "#CED38C", "mantaJ_SVfiltered" = "#DCC949", "smoove_SVfiltered" = "#BCA888", "cuteSV_SVfiltered" = "#CD8862", "sniffles_SVfiltered" = "#775B24")) +
    theme_light()
#dev.off()
```

## Overlapping SV calls
Of the six tools assessed here, we found significant variability in the number of SVs passing call quality thresholds. To examine the number of these passing calls that occur in two or more SV discovery tools, we ran SURVIVOR merge to identify SV intersecting within 50bp, 500bp and 1kb. Below we plot the number of SVs intersecting each of these comparisons.The above and below plots were restructured in Adobe Illustrator.  
```{r SV overlaps, echo=FALSE, fig.align='center'}
overlaps <-read.table("inputs/overlap_counts.tsv", sep = "\t", header = TRUE)
overlaps_50bp <- overlaps %>%
                filter(if_any(combination, ~grepl('_50bp' , .)))
#pdf("plots/Fig1_overlap_counts_full.pdf")
ggplot(overlaps_50bp, aes(x = reorder(combination, total), y = total)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  theme_light()
#dev.off()

#pdf("plots/Fig1_overlap_counts_ylim.pdf")
ggplot(overlaps_50bp, aes(x = reorder(combination, total), y = total)) +
  geom_bar(stat = "identity") +
  ylim(0,1300) +
  coord_flip() +
  theme_light()
#dev.off()
```

## Size Distributions

```{r size summaries, echo=FALSE, fig.align = 'center'}
filtered$size <- as.numeric(filtered$size)
filtered$logsize <- log(filtered$size)
head(filtered)
filtered %>%
  filter(if_any(data, ~grepl('smoove' , .))) %>%
  filter(if_any(SV_type, ~grepl('Duplication' , .))) %>%
  summary()
```
Here, we compared the size distribution of SVs called by all three tools. Due to the experimental set up violating many assumptions of independence, comparisons of variance between the tools were not made. Instead differences are in the context of means. Given the highly skewed size distribution across SV types in all data sets, the natural log of SV size was used to graph the size distribution and estimate the geometric mean.
```{r size distribution, echo=FALSE, fig.align='center'}
# Not transformed
ggplot(filtered, aes(x=size, fill=data)) +
  geom_density(alpha = 0.7) +
  xlim(0,10000) +
  labs(x = "Structural Variant Size", y = "Proportion", title = "Size Distribution of Filtered Structural Variants") +
  theme_light() +
  facet_wrap(~SV_type, scales = "free_y") +
  scale_fill_manual(values = c("delly_SVfiltered" = "#7D9D33", "mantaB_SVfiltered" = "#CED38C", "mantaJ_SVfiltered" = "#DCC949", "smoove_SVfiltered" = "#BCA888", "cuteSV_SVfiltered" = "#CD8862", "sniffles_SVfiltered" = "#775B24")) +
  theme_light()

ggplot(filtered, aes(x=size, fill=data)) +
  geom_histogram(alpha = 0.7, position = "dodge") +
  xlim(0,10000) +
  labs(x = "Structural Variant Size", y = "Proportion", title = "Size Distribution of Filtered Structural Variants") +
  theme_light() +
  facet_wrap(~SV_type, scales = "free_y") +
  scale_fill_manual(values = c("delly_SVfiltered" = "#7D9D33", "mantaB_SVfiltered" = "#CED38C", "mantaJ_SVfiltered" = "#DCC949", "smoove_SVfiltered" = "#BCA888", "cuteSV_SVfiltered" = "#CD8862", "sniffles_SVfiltered" = "#775B24")) +
  theme_light()

# Size transformed
#png("plots/Size_distributions.png")
ggplot(filtered, aes(x=logsize, fill=data)) +
  geom_density(alpha = 0.5) +
  labs(x = "Structural Variant Size", y = "Proportion", title = "Size Distribution of Filtered Structural Variants") +
  theme_light() +
  facet_wrap(~SV_type, scales = "free_y") +
  scale_fill_manual(values = c("delly_SVfiltered" = "#7D9D33", "mantaB_SVfiltered" = "#CED38C", "mantaJ_SVfiltered" = "#DCC949", "smoove_SVfiltered" = "#BCA888", "cuteSV_SVfiltered" = "#CD8862", "sniffles_SVfiltered" = "#775B24")) +
  theme_light()
#dev.off()

ggplot(filtered, aes(x=logsize, fill=data)) +
  geom_histogram(alpha = 0.7, position = "dodge") +
  labs(x = "Structural Variant Size", y = "Proportion", title = "Size Distribution of Filtered Structural Variants") +
  theme_light() +
  facet_wrap(~SV_type, scales = "free_y") +
  scale_fill_manual(values = c("delly_SVfiltered" = "#7D9D33", "mantaB_SVfiltered" = "#CED38C", "mantaJ_SVfiltered" = "#DCC949", "smoove_SVfiltered" = "#BCA888", "cuteSV_SVfiltered" = "#CD8862", "sniffles_SVfiltered" = "#775B24")) +
  theme_light()
```
## Chromosome Diversity
Variability in chromosome 'richness' SV counts and the relative proportion of each chromosome impacted by SVs were then assessed.

```{r Chromosome Comparisons, echo=FALSE, fig.align='center'}
filtered$proportion <- filtered$size / filtered$chrom_size

# The absolute count of unfiltered SVs per chromosome
ggplot(unfiltered, aes(x = chrom, fill = SV_type)) + 
  geom_histogram(stat = "count", position = "stack") +
  labs(x = "Chromosome", y = "Count", title = "Total Number of Structural Variants per Chromosome") +
  theme_light() + 
  facet_wrap(~data, scales = "free_y") +
  scale_fill_manual(values = c("Deletion" = "#7D9D33", "Duplication" = "#CD8862", "Insertion" = "#BCA888", "Inversion" = "#DCC949"))

# The absolute count of filtered SVs per chromosome
#pdf("plots/Fig2a_chr_distributions.pdf")
ggplot(filtered, aes(x = chrom, fill = SV_type)) + 
  geom_histogram(stat = "count", position = "stack") +
  labs(x = "Chromosome", y = "Count", title = "Number of Filtered Structural Variants per Chromosome") +
  theme_light() + 
  facet_wrap(~data, scales = "free_y") +
  scale_fill_manual(values = c("Deletion" = "#7D9D33", "Duplication" = "#CD8862", "Insertion" = "#BCA888", "Inversion" = "#DCC949"))
#dev.off()

# The proportion of the chromosome impacted by each filtered data set
pdf("plots/Fig2b_chr_distributions.pdf")
filtered %>%
  ggplot(aes(x = chrom, y = proportion, fill = SV_type)) +
  geom_bar(stat = "identity", position = "stack") +
  labs(x = "Chromosome", y = "Proportion of impacted base pairs", title = "Relative proportion of chromosome impacted by structural variants") +
  theme_light() +
  facet_wrap(~data, scales = "free_y") +
  ylim(0, 0.095) +
  scale_fill_manual(values = c("Deletion" = "#7D9D33", "Duplication" = "#CD8862", "Insertion" = "#BCA888", "Inversion" = "#DCC949"))
dev.off()
```

## Individual Richness
Here we explore the differences in the diversity of SV types across individual kākāpō.
```{r Individual Data, echo=FALSE, fig.align='center'}
delly_gen <- read.table("inputs/delly_generations.tsv", sep = "\t", header = TRUE)
#mantaB_gen <- read.table("inputs/mantaB_generations.tsv", sep = "\t", header = TRUE)
#mantaJ_gen <- read.table("inputs/mantaJ_generations.tsv", sep = "\t", header = TRUE)
smoove_gen <- read.table("inputs/smoove_generations.tsv", sep = "\t", header = TRUE)
cute_gen <- read.table("inputs/cute_generations.tsv", sep = "\t", header = TRUE)
sniff_gen <- read.table("inputs/sniffles_generations.tsv", sep = "\t", header = TRUE)
generations <- rbind(delly_gen, smoove_gen, cute_gen, sniff_gen)
```

As a start, we counted the number of SVs present within individual kākāpō, without considering genotype. That is to say, we counted presence/absence of SVs within individual kākāpō. This was done as reliably genotyping individuals can be challenging, and summary statistics for SVs (e.g., nucleotide diversity) have not yet been established.  
```{r Individual Richness, echo=FALSE, fig.align='center'}
pdf("plots/Fig4_indiv_richness.pdf")
ggplot(generations, aes(x=indiv, fill=SV_type)) +
  geom_histogram(alpha = 0.7, stat = 'count') +
  labs(x = "Individual", y = "Strucgtural Variant Count", title = "Number of SVs carried by individual kākāpō") +
  theme_light() +
  facet_wrap(~data, scales = "free_y") +
  scale_fill_manual(values = c("Deletion" = "#7D9D33", "Duplication" = "#CED38C", "Insertion" = "#DCC949", "Inversion" = "#BCA888")) +
  theme_light()
dev.off()
```
## Generational Diversity
Once we assessed the number of putative SVs carried by individual kākāpō in each of the six call sets, we examoined population-scale patterns of SV diversity. Given that the vast majority of the extant kākāpō population are descendended from a single population sourced from an off-shore island, we attempted to assess the diversity of SVs through the generations. 

Taking the generation data used to assess individual diversity above, we leveraged a meticulously maintained pedigree to identify individual kākāpō that are 'true' representatives of founders, F1 and F2 generations (i.e., not backcrossed).  
```{r Generational_Comparisons, echo=FALSE, fig.align='center'}
pdf("plots/Fig5_generation_counts.pdf")
generations %>%
  count(indiv, gen, data) %>%
  ggplot(aes(x=gen, y = n, fill=gen)) +
  geom_violin() +
  labs(x = "Generations", y = "Structural Variant Count", title = "Number of SVs carried by individual kākāpō", fill = "Generational Groupings") +
  theme_light() +
  facet_wrap(~data, scales = "free_y") +
  scale_fill_manual(values = c("RH0" = "#AD6B17","RH1" = "#CCB62F", "RH2" = "#D0C471", "SI0" = "#3E4331", "SI1" = "#66743B", "SI2" = "#BAC4C2")) +
  theme_light()
dev.off()
```

## Discriminant Analysis of Principal Components
Here we assess population structure across generations and kākāpō lineages. We leverage the genotype filtered SVs for the 117 individuals samples that represented founding, F1 and F2 generations in the Rakiura and Fiordland lineages.  
```{r load VCFs, echo=FALSE, fig.align='center'}
delly_vcf <- read.vcfR("inputs/pca/11_delly_genofilter_trio_target_samples.vcf")
delly_genind <- vcfR2genind(delly_vcf)
pop(delly_genind) <- as.factor(c("RH0", "RH2", "RH2", "RH1", "RH2", "RH1", "RH2", "RH1", "SI0", "SI1", "SI0", "SI1", "SI1", "SI2", "SI0", "SI0", "SI0", "SI0", "SI0", "SI0", "SI1", "SI1", "SI0", "SI0", "SI0", "SI1", "SI2", "SI1", "SI1", "SI1", "SI1", "SI1", "SI0", "SI0", "SI0", "SI1", "SI1", "SI0", "SI0", "SI1", "SI1", "SI1", "SI1", "SI1", "SI1", "SI1", "SI2", "SI1", "SI1", "SI1", "SI0", "SI1", "SI0", "SI0", "SI2", "SI1", "SI2", "SI1", "SI1", "SI1", "SI1", "SI0", "SI0", "SI0", "SI0", "SI0", "SI1", "SI1", "SI0", "SI1", "SI1", "SI0", "SI1", "SI1", "SI2", "SI1", "SI2", "SI0", "SI0", "SI0", "SI2", "SI1", "SI0", "SI1", "SI1", "SI1", "SI1", "SI1", "SI0", "SI1", "SI1", "SI1", "SI0", "SI0", "SI1", "SI0", "SI1", "SI0", "SI0", "SI1", "SI0", "SI0", "SI1", "SI1", "SI1", "SI1", "SI1", "SI1", "SI0", "SI1", "SI1", "SI2", "SI0", "SI1", "SI2", "SI1", "SI1"))

mantaB_vcf <- read.vcfR("inputs/pca/mantaB_target_samples.vcf")
mantaB_genind <- vcfR2genind(mantaB_vcf)
pop(mantaB_genind) <- delly_genind$pop

#mantaJ_vcf <- read.vcfR("inputs/pca/")
#mantaJ_genind <- vcfR2genind(mantaJ_vcf)
#pop(mantaJ_genind) <- delly_genind$pop

smoove_vcf <- read.vcfR("inputs/pca/06_smoove_genofilter_trio_target_samples.vcf")
smoove_genind <- vcfR2genind(smoove_vcf)
pop(smoove_genind) <- delly_genind$pop

cute_vcf <- read.vcfR("inputs/pca/08_cuteSV_genofilter_trio_target_samples.vcf")
cute_genind <- vcfR2genind(cute_vcf)
pop(cute_genind) <- delly_genind$pop

sniff_vcf <- read.vcfR("inputs/pca/08_sniffles_genofilter_trio_target_samples.vcf")
sniff_genind <- vcfR2genind(sniff_vcf)
pop(sniff_genind) <- delly_genind$pop
```

DAPCs were estimated for each of the six call sets. 
```{r DAPC, echo=FALSE, fig.align='center'}
myPal <- (c("#AD6D29", "#CCB732", "#CFC472", "#3F4432", "#67743C", "#BAC4C2" ))

pdf("plots/Fig6_delly_DAPC.pdf")
dapc.delly <- dapc(delly_genind, var.contrib = TRUE, scale = FALSE, n.pca = 35, n.da = nPop(delly_genind) -1)
scatter(dapc.delly, col=transp(myPal), scree.da = TRUE, scree.pca = TRUE, posi.da = "bottomleft", posi.pca = "topleft", cell = 1.5, cex = 2, cstar = 0)
100*dapc.delly$eig/sum(dapc.delly$eig)
assignplot(dapc.delly)
dev.off()

dapc.mantaB <- dapc(mantaB_genind, var.contrib = TRUE, scale = FALSE, n.pca = 40, n.da = nPop(mantaB_genind) -1)
scatter(dapc.mantaB, col = transp(myPal), scree.da = TRUE, scree.pca = TRUE, posi.da = "bottomleft", posi.pca = "topleft", cell = 1.5, cex = 2, cstar = 0)
perc.mantaB <- 100*dapc.mantaB$eig/sum(dapc.mantaB$eig)
assignplot(dapc.mantaB)

dapc.mantaJ <- dapc(mantaJ_genind, var.contrib = TRUE, scale = FALSE, n.pca = 40, n.da = nPop(mantaB_genind) -1)
scatter(dapc.mantaJ, col = transp(myPal), scree.da = TRUE, scree.pca = TRUE, posi.da = "bottomleft", posi.pca = "topleft", cell = 1.5, cex = 2, cstar = 0)
perc.mantaJ <- 100*dapc.mantaJ$eig/sum(dapc.mantaJ$eig)
assignplot(dapc.mantaJ)

pdf("plots/Fig6_smoove_DAPC.pdf")
dapc.smoove <- dapc(smoove_genind, var.contrib = TRUE, scale = FALSE, n.pca = 40, n.da = nPop(smoove_genind) -1)
scatter(dapc.smoove, col=transp(myPal), scree.da = TRUE, scree.pca = TRUE, posi.da = "bottomleft", posi.pca = "topleft", cell = 1.5, cex = 2, cstar = 0)
100*dapc.smoove$eig/sum(dapc.smoove$eig)
assignplot(dapc.smoove)
dev.off()

pdf("plots/Fig6_cuteSV_DAPC.pdf")
dapc.cute <- dapc(cute_genind, var.contrib = TRUE, scale = FALSE, n.pca = 20, n.da = nPop(cute_genind) -1)
scatter(dapc.cute, col=transp(myPal), scree.da = TRUE, scree.pca = TRUE, posi.da = "bottomleft", posi.pca = "topleft", cell = 1.5, cex = 2, cstar = 0)
100*dapc.cute$eig/sum(dapc.cute$eig)
assignplot(dapc.cute)
dev.off()

pdf("plots/Fig6_sniffles_DAPC.pdf")
dapc.sniff <- dapc(sniff_genind, var.contrib = TRUE, scale = FALSE, n.pca = 30, n.da = nPop(sniff_genind) -1)
scatter(dapc.sniff, col=transp(myPal), scree.da = TRUE, scree.pca = TRUE, posi.da = "bottomleft", posi.pca = "topleft", cell = 1.5, cex = 2, cstar = 0)
100*dapc.sniff$eig/sum(dapc.sniff$eig)
assignplot(dapc.sniff)
dev.off()
```